Content Picker refactoring into a single script

atuchin-m opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/40696

It's a part of work to remove Brave extension background page.
This PR rebuild content_picker feature to have a single script to inject and to all the work independently (without the background page).
There is no logic changes, it's mostly adopting the old code.

The next step is to inject the script via `CosmeticFiltersJSHandler`.

* the feature no more longer use a `chrome-extension://` iframe, it's replaced by a div + shadowDom 
* all external dependencies are declared in `api` object (currently it's still operates via extension API).
* `elementPicker.html` is fetched by the `content_element_picker.ts` and added to DOM dynamically;
* `elementPicker.js` code is merged with `content_element_picker.ts`, the file is deleted;
* `chrome.runtime` messaging is removed in favor of direct function calls; 

The old pipeline:
contentPicker iframe (elementPicker.html) <=> content_picker inject (`content_element_picker.ts`) <=> brave extension background page <=> browser process

The new pipeline:
content_picker inject <=> browser process

Security highlights:
* no `chrome-extension://` frame anymore (but IMHO it didn't give us any extra security, all the sensitive calls are still work thought `content_element_picker`.
* `shadowRoot.innerHTML = require('./elementPicker.html?raw')` is used to add the picker html to the page. `require` is resolved to the raw file content by webpack. We need this to make a single file to inject. Inlining html to js manually is a bad style (no lint/code completion/etc)
* `content_script` contexts is added `chrome.braveShields API` access. It's a trusted context and in fact the API are accessible indirectly (the background page proxies all the calls). The code will be removed in the next PR.
The only thing that can do the script in case it get highjacked is to add some adblock rules **for the loaded domain** (`domain##selector`)

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/content-picker-no-iframe/storybook.html)

antonok-edm reviewed (commented):
> The only thing that can do the script in case it get highjacked is to add some adblock rules for the loaded domain (`domain##selector`)

This doesn't seem like a great "feature" to have in an official build. My preference would be to include the next change in this PR as well, if possible.

atuchin-m commented on components/brave_extension/extension/brave_extension/content_element_picker.ts:
```diff
@@ -264,11 +276,17 @@ const quitElementPicker = () => {
 
-const launchElementPicker = () => {
-  pickerFrame = document.createElement('iframe')
+const attachElementPicker = async () => {
```
Hm, looks like after using webpack assets we don't need async anymore. I will remove it.

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/content-picker-no-iframe/linux/x64/storybook.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/content-picker-no-iframe/windows/x64/storybook.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/content-picker-no-iframe/macos/arm64/storybook.html)

atuchin-m commented:
>The only thing that can do the script in case it get highjacked is to add some adblock rules for the loaded domain (domain##selector)
@antonok-edm I'm just taking about the worst case scenario (the common thing of the old the old and new impl).
Nothing is actually changed in terms of security, because content_picker_element.ts context de-facto had the access to `braveShields.addSiteCosmeticFilter` via messaging to the background page

atuchin-m commented on components/brave_extension/extension/brave_extension/content_element_picker.ts:
```diff
@@ -264,11 +276,17 @@ const quitElementPicker = () => {
+  // Will be resolved by webpack to the file content.
+  // It's a trusted content so it's safe to use innerHTML.
+  // eslint-disable-next-line no-unsanitized/property
+  shadowRoot.innerHTML = require('./elementPicker.html?raw')
```
@petemill do you have any ideas?
We can just inline the whole HTML here, but it's super ugly.

atuchin-m replied:
@thypon Do you still believe that's not a good enough solution?
If we don't trust webpack processing there are possible sec issues in almost every browser js component.

petemill replied:
Well you could use Trusted Types to get a secure html string, but I'm not sure that would actually gain any restrictions here since all it's going to do is wrap it in a function which is basically saying "this is allowed". We'd still be passing in the same string from the build system which we trust.

So I could suggest how we can use the `html_to_wrapper` GN target template to wrap the HTML in a JS function which uses `static_types.ts`'s `static-types` policy and `createHTML` function, but I don't see that it adds anything beneficial here - if someone can think of a reason it would be helpful, I can outline the steps.

thypon replied:
this is mainly a provencance issue, I guess. We can use trusted types here, I feel.
We can have something like:

```javascript
const requireWithPolicy = trustedTypes.createPolicy("requirePolicy", {
  createHTML: function(requireString) {
    // check if the file is coming from webpack
    // check if the file is read-only
    // check if the permissions are correct for the file
    // TODO: implement the checks

    return require(requireString)
  },
});

let shadowRoot = /* ... */
const escaped = requireWithPolicy.createHTML("./elementPicker.html?raw")
console.log(escaped instanceof TrustedHTML) // true
el.innerHTML = escaped
```

atuchin-m replied:
1. does it give us any real extra protection?
2. how to implement these checks?
 
 BTW, There is no separate file, it's build-in in the bundle. here the bundle we inject: https://gist.github.com/atuchin-m/8b9c77fdddd29afd58528a1951025c1b

atuchin-m replied:
Discussed in DM.
The file content is just inlined in the resulting webpack js bundle, it's just a constant.
The only theoretical "weak" point here is `__webpack_require__`, but if it's hijacked, any loaded js module can be replaced, it doesn't create any new attack vectors.

fmarier commented:
> Nothing is actually changed in terms of security, because content_picker_element.ts context de-facto had the access to `braveShields.addSiteCosmeticFilter` via messaging to the background page

I thought we had fixed those issues:
- https://hackerone.com/reports/1255746
- https://hackerone.com/reports/1254125

atuchin-m commented:
> That would be only in the case of a compromised renderer, right? This is not normally accessible to the JS running on a website?

Yep, normally it's run in a isolated world and isn't accessible by the page js.
Although, if I'm not mistaken, chromium has a classification for such trusted contents and isolated worlds js is less trusted than chrome:// webUI.

I just want to highlight that maximum access level that this script has.
All the above is true for the current `master` and for the PR.

atuchin-m commented:
> thought we had fixed those issues

@fmarier 
I don't have access to hackerone.
Currently, the injected script can add rules for the current origin.
Call `chrome.runtime.sendMessage({type: 'cosmeticFilterCreate', selector: 'div'})` from Brave Extension context.
<img width="994" alt="image" src="https://github.com/user-attachments/assets/8df7f564-35da-4602-b076-0c6e62d8830e">
To be sure that the possible impact is limited to the current origin I've removed `host` params from `braveShields.addSiteCosmeticFilter`.

fmarier commented:
> I don't have access to hackerone.

Sorry about that, I should have linked to the two public issues:
- https://github.com/brave/brave-browser/issues/16883
- https://github.com/brave/brave-browser/issues/16863

> Currently, the injected script can add rules for the current origin.
> Call `chrome.runtime.sendMessage({type: 'cosmeticFilterCreate', selector: 'div'})` from Brave Extension context.

Yes, that was the [fix for the second issue](https://github.com/brave/brave-core/pull/17130).

fmarier commented:
> To be sure that the possible impact is limited to the current origin I've removed host params from braveShields.addSiteCosmeticFilter.

That sounds great. The high-level approach here looks fine to me. From what I can tell, this has the same security properties as before.

petemill commented on components/brave_extension/extension/brave_extension/content_element_picker.ts:
```diff
@@ -264,11 +275,17 @@ const quitElementPicker = () => {
   // This ensures a malicious page cannot modify the iframe contents.
-  pickerFrame.src = chrome.runtime.getURL('elementPicker.html')
+  pickerDiv = document.createElement('div')
+  const shadowRoot = pickerDiv.attachShadow({ mode: 'open' })
```
We could still use an iframe and inject to the iframe's document via `iframe.contentWindow.document`, but I don't think there's really any disadvantage/advantage between shadow DOM and iframe - if the document really wants to observe this and interfere then they could either way. Actually if you make this a 'closed' shadow, then the main world wouldn't be able to access the element's contents, and they can't monkey patch `attachShadow` in this isolated world, right? So if you change this to 'closed', then it could be more protected.

atuchin-m replied:
thanks for the suggestion, changed to "closed" shadow mode. 
BTW, `window` objects are unique in isolated world, so you can't just patch something (e.g. XHR request prototype) in the main world to affect isolated worlds.

petemill replied:
Yes but in a iframe or an open shadow, they could access the DOM elements and manipulate them. Less easy in a closed since they can't patch, so this is great.

petemill commented on components/brave_extension/extension/brave_extension/elementPicker.html:
```diff
@@ -1,130 +1,109 @@
+    overflow-y: auto;
+    overflow-wrap: break-word;
+  }
 
-    #rules-box {
-      height: 8em;
-      padding: 5px;
-      margin-bottom: 5px;
-      background-color: #fff;
-      border-radius: .1em;
-      overflow-y: auto;
-      overflow-wrap: break-word;
-    }
+  #rules-box>textarea {
+    height: 100%;
+    width: 100%;
+    resize: none;
+    border: none;
+    overflow: auto;
+  }
```
To fix the double parent + child scroll bars, remove padding and both overflow properties from this rule and add them to the textarea rule below. In addition, add box-sizing: border-box; to the textarea rule.

![image](https://github.com/user-attachments/assets/95119e3a-3988-43d4-b470-ee47479ced2f)


```
  #rules-box {
    height: 8em;
    margin-bottom: 5px;
    background-color: #fff;
    border-radius: .1em;
  }

  #rules-box>textarea {
    box-sizing: border-box;
    height: 100%;
    width: 100%;
    resize: none;
    border: none;
    padding: 5px;
    overflow: auto;
    overflow-wrap: break-word;
  }
```

<img width="394" alt="image" src="https://github.com/user-attachments/assets/dd6290dc-196e-42c6-94a0-3d2e11470a25">

atuchin-m replied:
done.
thanks for the suggestion.
